File: leetcode/algorithms/Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    if (s.size() == 0) {
      return 0;
    }
    unordered_map<char, int> cnts{};
    int i = 0, j = 0;
    int n = s.size();
    int res = 0;
    while (i < n) {
      while (j <= i && cnts[s[i]] > 0) {
        cnts[s[j]]--;
        j++;
      }
      cnts[s[i]]++;
      i++;
      res = max(res, i - j);
    }
    return res;
  }
};

File: leetcode/algorithms/Longest Substring Without Repeating Characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        lo, hi = 0, 0
        cnts = dict()
        res = 0
        while hi < len(s):
            if s[hi] not in cnts:
                cnts[s[hi]] = 0
            cnts[s[hi]] += 1
            if cnts[s[hi]] > 1:
                while cnts[s[hi]] > 1:
                    cnts[s[lo]] -= 1
                    lo += 1
            hi += 1
            res = max(res, hi - lo)
        return res
